### PR TITLE
Do not hang on unsupported HTTP methods

### PR DESCRIPTION
**Pull request by tlrx**
Unsupported HTTP methods are detected during requests dispatching which generates an appropriate error response. Sadly, this error is never sent back to the client because the method of the original request is checked again in `DefaultRestChannel` which throws again an `IllegalArgumentException` that is never handled.

This pull request changes the `DefaultRestChannel` so that the latest exception is swallowed, allowing the error message to be sent back to the client. It also eagerly adds the objects to close to the `toClose` list so that resources are more likely to be released  if something goes wrong during the response creation and sending.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/http/DefaultRestChannel.java [hunk 1]
```diff
@@ -171,4 +165,13 @@ private boolean isCloseConnection() {
     private boolean isHttp10() {
         return request.getHttpRequest().protocolVersion() == HttpRequest.HttpVersion.HTTP_1_0;
     }
+
+    // Determine if the request method is HEAD, swallowing any exception if the request method is invalid
+    private boolean isHead() {
+        try {
+            return RestRequest.Method.HEAD == request.method();
+        } catch (Exception ignored) {
+            return false;
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/http/DefaultRestChannel.java]: Maybe pass the response into this method and assert that it's an error response via its status? (I think that would make it easier to understand why this is ok to swallow). Technically I guess this should have status `405` (method not allowed?)

File: server/src/test/java/org/elasticsearch/http/DefaultRestChannelTests.java [hunk 2]
```diff
@@ -442,4 +510,41 @@ public RestStatus status() {
             return RestStatus.OK;
         }
     }
+
+    private class TestBytesReference extends BytesReference implements Releasable {
```
Comment (original-brownbear) [inline server/src/test/java/org/elasticsearch/http/DefaultRestChannelTests.java]: I think you could shorten this by just using `org.elasticsearch.common.bytes.ReleasablePagedBytesReference` with an empty bytes array and passing your close logic as the `Releasable` in its constructor?

**Review by original-brownbear — COMMENTED** [review]
I left two suggestions. I wonder if we should put some effort in here to actually return a `405` response?

File: server/src/test/java/org/elasticsearch/http/DefaultRestChannelTests.java (hunk 2 above)
Comment (tlrx) [reply server/src/test/java/org/elasticsearch/http/DefaultRestChannelTests.java]: I pushed 95efb064e18e85223160854dfcb70d068b277499

File: server/src/main/java/org/elasticsearch/http/DefaultRestChannel.java (hunk 1 above)
Comment (tlrx) [reply server/src/main/java/org/elasticsearch/http/DefaultRestChannel.java]: I pushed a64c1896d4695d4ec28cec57044414435d28596d

**Comment by tlrx** [comment]
Thanks @original-brownbear. I followed your comments and beefed up a bit more how unsupported HTTP methods are handled. Also for favicon.ico as it has a special short path.

File: server/src/main/java/org/elasticsearch/rest/RestController.java [hunk 3]
```diff
@@ -247,20 +246,29 @@ boolean dispatchRequest(final RestRequest request, final RestChannel channel, fi
...
+                    requestHandled = true;
+                } else if (validMethodSet.contains(requestMethod) == false
+                    && (requestMethod == RestRequest.Method.OPTIONS)) {
+                    handleOptionsRequest(request, channel, validMethodSet);
+                    requestHandled = true;
+                } else {
+                    requestHandled = false;
+                }
+            } catch (IllegalArgumentException e) {
+                handleUnsupportedHttpMethod(request, channel, validMethodSet, e);
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/rest/RestController.java]: More of an optional question/suggestion: Can't we just do this a lot simpler by simply invoking this method upstream in `org.elasticsearch.rest.RestController#tryAllHandlers` even if it means that we have to duplicate the call to `getValidHandlerMethodSet`? (we'd save the whole change in this method and one level of indent)

**Review by original-brownbear — APPROVED** [review]
LGTM, left one optional suggestion to save one round of catching the `IllegalArgumentException`

File: server/src/main/java/org/elasticsearch/rest/RestController.java (hunk 3 above)
Comment (tlrx) [reply server/src/main/java/org/elasticsearch/rest/RestController.java]: I pushed eac0baf906e04ec87f82d96224d1507a52b33c68, what do you think?

Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/rest/RestController.java]: Yea I like this a lot better :)

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/1
Jenkins run elasticsearch-ci/default-distro
Jenkins run elasticsearch-ci/bwc

(just random timeouts it seems)

File: server/src/main/java/org/elasticsearch/rest/RestController.java (hunk 3 above)
Comment (tlrx) [reply server/src/main/java/org/elasticsearch/rest/RestController.java]: Thanks!

**Comment by tlrx** [comment]
@elasticmachine update branch

**Comment by tlrx** [comment]
Thanks @original-brownbear

**Pull request merged**